### DISCUSSION TITLE
Issues with serialtest.py - Have troubleshooted

### DISCUSSION THREAD

**Discussion by ethanlchristensen** [jgyates/genmon#1121] [Questions] [answered]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).

### Version

1.19.02

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have not submitted logs

### What is/are your question(s)?

I am having issue with getting the serialtest.py to work. I have walked through many of the trouble shooting steps, but can't seem to get it to work out. Was seeing if I could get some eyes from the community. I figured I would share here some output information.

```bash
mayor@mayor:~/genmon/OtherApps $ sudo python serialconfig.py -c

Using files: /boot/firmware/config.txt and /boot/firmware/cmdline.txt

Model: Raspberry Pi Zero 2 W Rev 1.0

Checking : Is enable UART in boot config: OK
Checking : Is BT overlay disabled: Fail
Checking : Serial console command line removed: OK
Checking : Serial console service disabled: OK
Checking : BT service disabled: OK

Serial port may not work as expected. Not all required settings changes were detected.

mayor@mayor:~/genmon/OtherApps $  pinctrl get 14,15
14: a0    -- | hi // GPIO14 = TXD0
15: a0    -- | hi // GPIO15 = RXD0
mayor@mayor:~/genmon/OtherApps $ dmesg | grep tty
[    0.000000] Kernel command line: coherent_pool=1M 8250.nr_uarts=1 snd_bcm2835.enable_headphones=0 snd_bcm2835.enable_hdmi=1 snd_bcm2835.enable_hdmi=0  smsc95xx.macaddr=B8:27:EB:7B:0B:1C vc_mem.mem_base=0x1ec00000 vc_mem.mem_size=0x20000000  console=tty1 root=PARTUUID=a10fa7f1-02 rootfstype=ext4 fsck.repair=yes rootwait cfg80211.ieee80211_regdom=US
[    0.000268] printk: console [tty1] enabled
[    2.151349] 3f215040.serial: ttyS0 at MMIO 0x3f215040 (irq = 71, base_baud = 50000000) is a 16550
[    2.153575] serial serial0: tty port ttyS0 registered
[    2.239740] 3f201000.serial: ttyAMA0 at MMIO 0x3f201000 (irq = 99, base_baud = 0) is a PL011 rev2
[    6.147642] systemd[1]: Created slice system-getty.slice - Slice /system/getty.
mayor@mayor:~/genmon/OtherApps $
mayor@mayor:~/genmon/OtherApps $ sudo python serialtest.py

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5
Serial port opened
write data: sent test string
waiting to received data....
FAILED: Sent data does not match receive. Received 0 bytes
```

**Comment by jgyates** [comment]
@ethanlchristensen 

A few comments:

1) You should be using the latest version which is 1.19.03
2) You did not mention what operating system and version you are using. This can be relevant. 
3) You are testing /dev/serial0 but you do not show if the symbolic links were setup properly. To validate this please show the output of  this command:

      ls -la /dev/ser*

4) You are showing the output of the serialconfig command with the -c option to check the config. What command did you use to set the config (the -e option) or did you modify the config files manually?

**Comment by ethanlchristensen** [comment]
Apologies for that @jgyates 

This was configured with the serialconfig.py -e method

Here is some of this missing information:
```bash
mayor@mayor:~/genmon/OtherApps $ ls -la /dev/ser*
lrwxrwxrwx 1 root root 7 May 24 16:17 /dev/serial0 -> ttyAMA0
mayor@mayor:~/genmon/OtherApps $ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

**Reply by jgyates** [reply, marked as answer]
You do not mention how you serial cabling currently configured? Is there a loopback between pin 14 and 15? Is there a RS-232 converter installed? I would do the test without the RS-232 converter and a loopback on pins 14 and 15. You can send a photo of your loopback for sanity checking if you like

Nothing so far look out of the ordinary except your timout errors. The software appears to think that it is transmitting but nothing is received. If you have a loopback connected to pin 14 and 15 then you can try this;

1) restore you config files with serailconfig.py

    sudo python3 serialconfig.py -r

2) reboot

     sudo reboot now

3) Set the serial config with bluetooth enabled

    sudo python3 serialconfig.py -e -b

4) reboot

5) test the serial port with the loopback, with genmon not loaded, using serialtest.py

If that does not work then either your loopback is bad or the pi zero is defective.

**Reply by ethanlchristensen** [reply]
OH! This passed over my head that I needed to short 14 and 15 together, I thought have the HAT was enough. Shorting 14 and 15 == success. Sorry for the bother, and thanks for the help. [tips hat]
```bash
mayor@mayor:~/genmon/OtherApps $ sudo python3 serialtest.py

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5
Serial port opened
write data: sent test string
waiting to received data....
PASSED! Loopback successful
```

**Reply by ethanlchristensen** [reply]
![image](https://github.com/jgyates/genmon/assets/55725575/25ad2292-4f39-4bba-bc34-bd48952fe305)

**Reply by ethanlchristensen** [reply]
The only other issue I had was the molex cable not powering the PI. I have an evolution 1 generator with only the Accessory port 1. Accessory port 2 is missing / not there.

**Reply by jgyates** [reply]
I am glad it is working for you. 

the Evo 1 and Evo 2 both have the 8 pin molex. The Evo 1 and 2 have 5V and ground. the Evo2 has 12V

https://github.com/jgyates/genmon/wiki/3.1--Making-a-Cable

some models do not label them the same way. If there is an 8 pin molex, that has TX, RX, 5V and ground. Having said that, the battery is probably a better method to power the pi using the voltage converter listed in the wiki.

**Reply by ethanlchristensen** [reply]
![image](https://github.com/jgyates/genmon/assets/55725575/d90ef13b-c9d4-4a87-aae3-3b826f5bda47)

Is this that converter? I had ordered a preloaded kit from PrintSizeMe.

**Reply by jgyates** [reply]
That's it.

thanks for the donation!

**Reply by ethanlchristensen** [reply]
No worries, seems like we are all set. Your work is worth more than that donation 😁. Well maintained and very knowledgeable!
